A test and a comment to verify BRIDGE-1350 is fixed.

alxdarksage opened the pull request:
Changes to address recent failures in production:
- ParticipantService changed so participant is always returned with health code. On error was calling getParticipant() and getting a participant record with a health code that was being saved in the session. Instead for the appropriate method, we remove healthCode if the study has disabled healthCode export.
- null pointer precondition check for healthCode before setting session in cache;
- when initializing session, if a session exists, keep the session tokens but update the session. Another issue was a session being set from the verifyEmail call which comes from a web page, not the client application, and lacks the User-Agent header to properly construct this session. consenting to research call fixes the session now.
- always update session cookie when setting the session in the cache. Not known to cause a problem but possibly only because no one looks at the cookie for this.
- fixed code path when creating accounts that may have been creating a healthCode and then throwing it away rather than saving it. Probably did not cause an issue since on sign in/consent it would be fixed in the course of constructing an account. But still.
- improved cache provider test based on examination of real session in production
- refactored tests as appropriate (e.g. moved the check that health code was not being included from the service to the controller).

alxdarksage committed:
A test and a comment to verify BRIDGE-1350 is fixed.

alxdarksage committed:
Merging. I'm pretty confused right now though why these changes don't show in PR to upstream.
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into delete-user

Conflicts:
	app/org/sagebionetworks/bridge/services/AuthenticationService.java
	test/org/sagebionetworks/bridge/services/AuthenticationServiceTest.java

DwayneJengSage commented:
Does this change re-apply https://github.com/Sage-Bionetworks/BridgePF/pull/1134 ?

Would it be cleaner to re-apply that pull request in a separate pull request, and then make a separate pull request with the fix on top of that? This would make reviewing the fix much much much easier.

DwayneJengSage commented:
While the new code was deployed to prod yesterday, we were seeing errors that suggest either the participant was missing from the session or the health code was missing from the participant. Was this bug fixed?

This is critical to fix, as it prevents users from uploading.

DwayneJengSage commented:
Filing https://sagebionetworks.jira.com/browse/BRIDGE-1353 as a blocker on this pull request

alxdarksage committed:
- ParticipantService only returns StudyParticipant with a healthCode, and it is removed in controller if no permissions to see it
- other small fixes based on examination of the code yesterday, such as no longer refreshing user's session after updating their StudyParticipant record (without information from the user's request, this can break their consent status records, they must be signed out to force them to sign in and pick up their new session state).

alxdarksage committed:
Augment a test to verify we are picking up fields like healthCode and they are in the session.

alxdarksage committed:
With comment.

alxdarksage committed:
Adding a precondition check for healthCode before storing session in cache.

alxdarksage commented on test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoMultipleSubpopulationsTest.java:
```diff
@@ -1,109 +0,0 @@
-
-@ContextConfiguration("classpath:test-context.xml")
-@RunWith(SpringJUnit4ClassRunner.class)
-public class StormpathAccountDaoMultipleSubpopulationsTest {
```
Test here was moved to a different class, still exists (see below).

DwayneJengSage replied:
Actually, this specific change reverts the fix for ConsentServiceTest. I know for a fact that the version of this test in StormpathAccountDaoTest will break ConsentServiceTest, which is why it was moved to its own file in the first place.

DwayneJengSage replied:
Restore the entire file.

alxdarksage replied:
restored

DwayneJengSage commented on test/org/sagebionetworks/bridge/dynamodb/DynamoExternalIdDaoTest.java:
```diff
@@ -4,15 +4,15 @@
 import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
```
This change undoes the ConsentServiceTest fix. Revert the entire file.

alxdarksage replied:
Should revert on update

DwayneJengSage commented on test/org/sagebionetworks/bridge/dynamodb/DynamoParticipantOptionsDaoMockTest.java:
```diff
@@ -1,28 +0,0 @@
-
-import org.sagebionetworks.bridge.TestConstants;
-
-public class DynamoParticipantOptionsDaoMockTest {
```
This undoes the ConsentServiceTest fix. Restore the entire file.

alxdarksage replied:
Should be restored on update

DwayneJengSage commented on test/org/sagebionetworks/bridge/dynamodb/DynamoParticipantOptionsDaoTest.java:
```diff
@@ -178,6 +183,28 @@ public void updateAllOptions() {
     }
+    
+    @Test
+    public void updateNoOptions() {
```
This change undoes the ConsentServiceTest fix. Revert the entire file.

alxdarksage replied:
Should revert on update

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java:
```diff
@@ -159,17 +159,30 @@ public void oddParametersUseDefaults() throws Exception {
-        StudyParticipant studyParticipant = new StudyParticipant.Builder().withFirstName("Test").build();
-        
-        when(participantService.getParticipant(STUDY, CALLER_ROLES, ID)).thenReturn(studyParticipant);
+        STUDY.setHealthCodeExportEnabled(true);
```
Modifying a static member will persist that change across tests. This is bad. If you need to modify the study, create a new study for each test instead of modifying the old one.

alxdarksage replied:
Converted to instance property.

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java:
```diff
@@ -159,17 +159,30 @@ public void oddParametersUseDefaults() throws Exception {
+    
+    @Test
+    public void getParticipantWithNoHealthCode() throws Exception {
+        STUDY.setHealthCodeExportEnabled(false);
```
Same here.

alxdarksage replied:
Changed everywhere

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/AuthenticationServiceTest.java:
```diff
@@ -319,12 +320,12 @@ public void invalidDataGroupsAreRejected() throws Exception {
-        verify(authServiceSpy).requestResetPassword(any(Study.class), any(Email.class));
+        authService.signUp(testUser.getStudy(), testUser.getStudyParticipant());
+        verify(authService).requestResetPassword(any(Study.class), any(Email.class));
     }
```
This modifies member variable authService. Past experience tells me that JUnit and TestNG don't reset member variables between tests. This could cause corruption across multiple tests. Revert this test to use a local variable.

alxdarksage replied:
reverted

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/HealthCodeServiceMockTest.java:
```diff
@@ -1,45 +0,0 @@
-import org.sagebionetworks.bridge.dao.HealthIdDao;
-import org.sagebionetworks.bridge.models.accounts.HealthId;
-
-public class HealthCodeServiceMockTest {
```
This reverts the ConsentServiceTest fix. Restore the entire file.

alxdarksage replied:
Added back

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/HealthCodeServiceTest.java:
```diff
@@ -24,6 +34,16 @@
+    @After
+    public void after() {
+        clearDynamo();
+    }
```
There is no need to clear dynamo before and after each test run. The test runs fine with other entries in the table.

Clearing Dynamo means our local test accounts lose their healthcodes, which makes local testing a pain.

alxdarksage replied:
removed (this must have been removed by your consent services changes, then added back)

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/HealthCodeServiceTest.java:
```diff
@@ -34,4 +54,41 @@ public void createAndRetrieveHealthId() {
+        HealthId id2 = healthCodeService.getMapping("456");
+        assertNull(id2);
+        
+    }
```
This reverts the ConsentServiceTest fix. Revert the entire file.

alxdarksage replied:
reverted

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/StudyServiceTest.java:
```diff
@@ -56,18 +55,15 @@
-        studyService.setCacheProvider(mockCache);
+        cache = mock(CacheProvider.class);
+        studyService.setCacheProvider(cache);
     }
```
This reverts the ConsentServiceTest fix. Revert the entire file.

alxdarksage replied:
reverted

DwayneJengSage commented on test/org/sagebionetworks/bridge/stormpath/StormpathAccountDaoTest.java:
```diff
@@ -307,7 +310,62 @@ public void canResendEmailVerification() throws Exception {
     }
+    
+    @Test
+    public void canSetAndRetrieveConsentsForMultipleSubpopulations() {
```
This test breaks ConsentServiceTest. Revert the entire file.

alxdarksage replied:
removed

DwayneJengSage commented on app/org/sagebionetworks/bridge/play/controllers/ParticipantController.java:
```diff
@@ -43,7 +43,7 @@ public Result getSelfParticipant() {
         Study study = studyService.getStudy(session.getStudyIdentifier());
         
-        StudyParticipant participant = participantService.getParticipant(study, NO_CALLER_ROLES, session.getId());
+        StudyParticipant participant = participantService.getParticipant(study, session.getId(), false);
```
This exposes health code to end users. Should we filter here?

alxdarksage replied:
It's marked @JsonIgnore, however, the encryptedHealthCode is being exposed. Let's get this PR fixed and merged and I'll create a follow-up PR to hide encryptedHealthCode in the API.

alxdarksage replied:
For now I'm removing healthCode and encryptedHealthCode from the JSON just to be sure this happens. This can be replaced with a Jackson filter in a follow-on update.

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java:
```diff
@@ -257,11 +269,11 @@ public void updateParticipantRequiresIdMatch() throws Exception {
         Result result = controller.getSelfParticipant();
         
-        verify(participantService).getParticipant(STUDY, NO_CALLER_ROLES, ID);
+        verify(participantService).getParticipant(STUDY, ID, false);
```
Do we need a check to make sure we're filtering out health code?

alxdarksage replied:
Added

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java:
```diff
@@ -273,7 +285,7 @@ public void getSelfParticipant() throws Exception {
         // All values should be copied over here.
         StudyParticipant participant = TestUtils.getStudyParticipant(ParticipantControllerTest.class);
-        doReturn(participant).when(participantService).getParticipant(STUDY, NO_CALLER_ROLES, ID);
+        doReturn(participant).when(participantService).getParticipant(STUDY, ID, false);
```
Add a check to verify that healthcode is being saved into the session.

alxdarksage replied:
Added

DwayneJengSage commented on app/org/sagebionetworks/bridge/play/controllers/ParticipantController.java:
```diff
@@ -115,12 +120,6 @@ public Result updateParticipant(String userId) {
-        CriteriaContext context = new CriteriaContext.Builder()
-                .withStudyIdentifier(study.getStudyIdentifier()).build();
-        session = authenticationService.updateSession(study, context, userId);
-        updateSession(session);
```
Do we no longer need to update the session?

alxdarksage replied:
Yes, this can cause errors because we're updating another user's session when we don't know their app version, and this could be used to determine consent status. I would rather that administrators purge the user's session if they want the user to pick up the changes, and that will force the user to sign in again.

alxdarksage replied:
We could automate this. Currently though there's a button and an API to do it as a separate step. If you sign people out, you have to do work on the client to determine if this was an edit of the current user's session, etc, which is error-prone.

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/UserProfileControllerTest.java:
```diff
@@ -158,7 +158,7 @@ public void getUserProfile() throws Exception {
     public void updateUserProfile() throws Exception {
         StudyParticipant participant = new StudyParticipant.Builder().withExternalId("originalId").build();
-        doReturn(participant).when(participantService).getParticipant(study, Sets.newHashSet(), ID);
+        doReturn(participant).when(participantService).getParticipant(study, ID, false);
```
Add check to make sure healthCode was added to session?

alxdarksage replied:
Added

DwayneJengSage commented on test/org/sagebionetworks/bridge/services/ParticipantServiceTest.java:
```diff
@@ -402,8 +401,8 @@ public void getStudyParticipant() {
-        participant = participantService.getParticipant(STUDY, Sets.newHashSet(DEVELOPER), ID);
-        assertNull(participant.getHealthCode());
+        participant = participantService.getParticipant(STUDY, ID, false);
+        assertEquals(HEALTH_CODE, participant.getHealthCode());
```
Minor: Comment is not accurate, should say something about always getting healthcodes regardless of history.

alxdarksage replied:
Removed. This tests the same thing as above and was originally because roles were being switched in the test, which is no longer the case.

DwayneJengSage commented on app/org/sagebionetworks/bridge/stormpath/StormpathAccountDao.java:
```diff
@@ -392,9 +392,11 @@ private Account constructAccount(StudyIdentifier studyId, com.stormpath.sdk.acco
         if (healthId == null) {
             healthId = healthCodeService.createMapping(studyId);
+            account.setHealthId(healthId);
             updateAccount(account);
```
Add unit test to ensure we're updating the account with a healthId?

alxdarksage replied:
StormpathAccountDaoMockTest.getAccountCreatesHealthCode() tests this. If you remove that setHealthId now that it's in an if/else statement, the test breaks. If you comment out the other branch, the authenticate test fails because it verifies that we initialize an existing healthCode with a healthId.

DwayneJengSage replied:
getAccountCreatesHealthCode() only tests that createMapping() was called. It doesn't test that setHealthId() was called before updateAccount(). We should have a test to exercise this bug fix.

alxdarksage replied:
added verification of customData being called with ID and then saved

DwayneJengSage commented on test/org/sagebionetworks/bridge/cache/CacheProviderTest.java:
```diff
@@ -306,12 +309,13 @@ private void assertSession(String json) {
+        assertEquals("ABC", participant.getExternalId());
         
-        SubpopulationGuid apiGuid = SubpopulationGuid.create("api");
+        Encryptor ENCRYPTOR = new AesGcmEncryptor(BridgeConfigFactory.getConfig().getProperty("bridge.healthcode.redis.key"));
```
We can make the encryptor a static member var.

alxdarksage replied:
moved

DwayneJengSage commented:
Done reviewing. It also looks like you need to resolve some merge conflicts.

alxdarksage committed:
- reverting a bunch of stuff I undid in Dwayne's commit
- updates based on code review
Merge branch 'develop' of https://github.com/Sage-Bionetworks/BridgePF into delete-user

Conflicts:
	test/org/sagebionetworks/bridge/dynamodb/DynamoScheduledActivityDaoTest.java
	test/org/sagebionetworks/bridge/services/AuthenticationServiceTest.java

alxdarksage committed:
Fixing whitespace causing file diffs.

alxdarksage commented on app/org/sagebionetworks/bridge/play/controllers/ParticipantController.java:
```diff
@@ -100,8 +98,14 @@ public Result getParticipant(String userId) {
+            participant = new StudyParticipant.Builder().copyOf(participant)
+                    .withHealthCode(null).build();
+        }
+        
```
For now I'm removing healthCode and encryptedHealthCode from the JSON just to be sure this happens. This can be replaced with a Jackson filter in a follow-on update.

alxdarksage committed:
Changes to reduce the diff of the PR, and to clarify/document code changes.

alxdarksage committed:
Removing another whitespace diff

alxdarksage committed:
Fixing tests now that we're vacating healthCode/encryptHealthCode in ParticipantController (for now).

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java:
```diff
@@ -257,23 +272,28 @@ public void updateParticipantRequiresIdMatch() throws Exception {
                 .readValue(Helpers.contentAsString(result), StudyParticipant.class);
 
         assertEquals("Test", deserParticipant.getFirstName());
+        assertNull(deserParticipant.getHealthCode());
```
This is a vacuous test. The participant we return from the mock already doesn't contain the health code, so we're not exercising the code that filters out health code.

DwayneJengSage replied:
Since we explicitly clear out both health code and encrypted health code, we should also explicitly check both health code and encrypted health code here.

alxdarksage replied:
I've changed these so both health code and encrypted health code are in the participant returned from the mock, and both are verified to be absent as returned from the controller.

DwayneJengSage commented:
It looks like github swallowed one of my comments. On StormpathAccountDao line 395 "getAccountCreatesHealthCode() only tests that createMapping() was called. It doesn't test that setHealthId() was called before updateAccount(). We should have a test to exercise this bug fix."

DwayneJengSage commented on app/org/sagebionetworks/bridge/play/controllers/ParticipantController.java:
```diff
@@ -45,7 +45,10 @@ public Result getSelfParticipant() {
-        return okResult(participant);
+        // Do not return healthCode. This will shortly be handled with Jackson filters.
+        StudyParticipant updated = new StudyParticipant.Builder().copyOf(participant)
+                .withHealthCode(null).withEncryptedHealthCode(null).build();
```
Is it necessary to clear out both health code and encrypted health code? They're both backed by the same member var. Or is this a "just in case" kind of thing, in case we need to change the implementation?

alxdarksage replied:
This is a temporary "just in case", I'm going to replace this with Jackson filters which is a little cleaner.

DwayneJengSage commented on test/org/sagebionetworks/bridge/play/controllers/ParticipantControllerTest.java:
```diff
@@ -146,42 +149,55 @@ public void getParticipants() throws Exception {
+        StudyParticipant retrievedParticipant = BridgeObjectMapper.get().readValue(json, StudyParticipant.class);
+        
+        assertEquals("Test", retrievedParticipant.getFirstName());
+        assertNull(retrievedParticipant.getHealthCode());
```
Similarly, do we want to explicitly test both getHealthCode() and getEncryptedHealthCode()?

alxdarksage replied:
Added this to the test (yes we want to test this until we can change it).

alxdarksage committed:
Updates based on code review.

alxdarksage committed:
Reducing test verbosity, formatting JSON.
